src/hotspot/share/classfile/protectionDomainCache.hpp

```diff
@@ -29,7 +29,8 @@
 #include "oops/weakHandle.hpp"
 #include "runtime/atomic.hpp"
 
-// The ProtectionDomainCacheTable contains all protection domain oops as WeakHandles.
+// The ProtectionDomainCacheTable maps all java.security.ProtectionDomain objects that are
+// registered by DictionaryEntry::add_protection_domain() to a unique WeakHandle.
```

Comment:
Now that I understand what this table does, this comment is confusing to me. The table maps each PD to itself (using the WeakHandle as the actual key and value) as a means to track which PDs have been seen/registered. I would describe it thus:
```
// The ProtectionDomainCacheTable records all of the java.security.ProtectionDomain instances that have
// been registered by DictionaryEntry::add_protection_domain(). We use a hashtable to
map each  PD 
// instance to itself (using WeakHandles) to allow for fast lookup.
```

Reply:
“Mapping to itself” is how this table works internally. I am not sure if this information is useful here. But the purpose of this table is really to keep track of all PDs that have been registered and not yet garbage collected.

Reply:
"mapping to itself" is more useful than "mapping ... to a unique Weakhandle" - which is even more of an internal implementation detail. I found the use of this table very hard to discern based on the internal use of the hashtable, as there is no real mapping operation - we simply track if a PD has been seen or not. The use of the hashtable is purely for lookup convenience - we could instead have a linked-list of PD's that we traverse for lookup.
So perhaps we drop my second sentence above, and move it to where the hashtable itself is declared?

Reply:
"mapping a PD to a unique Weakhandle" is not an implementation detail. It's the only useful API provided by this class:

```
WeakHandle obj = ProtectionDomainCacheTable::add_if_absent(protection_domain);
```

and that's the reason I question why this table is needed at all.

Reply:
The fact it is a weakhandle is an implementation detail. The table simply records whether a PD (wrapped in a WeakHandle) has been seen.

Reply:
Since I know what this table does, either comment is fine if it helps someone understand it.
The "map each PD to itself" is pretty odd to me too.   How about collect each PD for fast lookup in a hashtable?  The code says how it's mapped.  There isn't that much code and it's easy to see how the Key is mapped.

Reply:
// The ProtectionDomainCacheTable maps all java.security.ProtectionDomain objects that are
// registered by DictionaryEntry::add_protection_domain() to a unique entry.  The entry
// is a WeakHandle that holds the protection domain oop.

or points to....  either is accurate.

Reply:
It is the "maps ... to a unique entry" that I find most problematic - it begs the question as to what the unique entry is, when in reality it maps a PD instance (wrapped in a WeakHandle) to itself (wrapped in a WeakHandle).

For the sake of progress, approved.

Reply:
The "unique" means there's only one value in the table for this protection domain oop.  So the oop is unique.  The use of word "mapping" might be what's confusing.
If we ever have to visit this code again (which I hope not, except to remove it because the security manager is finally removed since not many use it), we can see if this comment still makes sense to at least some of us.
Thanks for approving for progress.